ArcGISRuntimeSDKQt_CppSamples/UtilityNetwork/ValidateUtilityNetworkTopology/ValidateUtilityNetworkTopology.cpp

```diff
@@ -0,0 +1,671 @@
+#include "MapQuickView.h"
+#include "MapTypes.h"
+#include "PictureMarkerSymbol.h"
+#include "Point.h"
+#include "Portal.h"
+#include "PortalItem.h"
+#include "QueryParameters.h"
+#include "ServiceFeatureTable.h"
+#include "ServiceGeodatabase.h"
+#include "ServiceTypes.h"
+#include "ServiceVersionInfo.h"
+#include "ServiceVersionParameters.h"
+#include "SimpleLabelExpression.h"
+#include "SimpleMarkerSymbol.h"
+#include "SpatialReference.h"
+#include "SubtypeSublayer.h"
+#include "SymbolTypes.h"
+#include "TaskTypes.h"
+#include "TextSymbol.h"
+#include "UtilityAssetGroup.h"
+#include "UtilityAssetType.h"
+#include "UtilityDomainNetwork.h"
+#include "UtilityElement.h"
+#include "UtilityElementTraceResult.h"
+#include "UtilityNetwork.h"
+#include "UtilityNetworkCapabilities.h"
+#include "UtilityNetworkDefinition.h"
+#include "UtilityNetworkListModel.h"
+#include "UtilityNetworkSource.h"
+#include "UtilityNetworkState.h"
+#include "UtilityNetworkTypes.h"
+#include "UtilityNetworkValidationJob.h"
+#include "UtilityTerminal.h"
+#include "UtilityTerminalConfiguration.h"
+#include "UtilityTier.h"
+#include "UtilityTraceParameters.h"
+#include "UtilityTraceResult.h"
+#include "UtilityTraceResultListModel.h"
+#include "UtilityValidationResult.h"
+#include "ValidateUtilityNetworkTopology.h"
+#include "Viewpoint.h"
+
+#include <QUuid>
+#include <QFuture>
+
+using namespace Esri::ArcGISRuntime;
+
+
+ValidateUtilityNetworkTopology::ValidateUtilityNetworkTopology(QObject* parent /* = nullptr */) :
+  QObject(parent),
+  m_cred(new Credential("editor01", "S7#i2LWmYH75", this))
+{
+  m_portal = new Portal(QUrl("https://sampleserver7.arcgisonline.com/portal/sharing/rest"), m_cred, this);
+
+  m_portalItem = new PortalItem(m_portal, "6e3fc6db3d0b4e6589eb4097eb3e5b9b", this);
+
+  m_map = new Map(m_portalItem, this);
+
+  connect(m_map, &Map::doneLoading, this, [this]()
+  {
+    // Load and switch utility network version
+    m_message = QString("Loading the utility network...");
+    emit messageChanged();
+
+    m_utilityNetwork = m_map->utilityNetworks()->first();
+    m_utilityNetwork->load();
+
+    m_serviceGeodatabase = m_utilityNetwork->serviceGeodatabase();
+    const QString uuid = QUuid().createUuid().toString(QUuid::WithoutBraces);
+
+    // Restrict editing and tracing on a random branch
+    ServiceVersionParameters* params = new ServiceVersionParameters(this);
+    params->setName("ValidateNetworkTopology_"+uuid);
+    params->setAccess(VersionAccess::Private);
+    params->setDescription("Validate network topology with ArcGIS Runtime");
+
+    connect(m_serviceGeodatabase, &ServiceGeodatabase::doneLoading, this, [this, params]()
+    {
+      if (m_serviceGeodatabase->loadStatus() == LoadStatus::Loaded)
+      {
+        m_serviceGeodatabase->createVersionAsync(params).then(this, [this](ServiceVersionInfo* serviceVersionInfo)
+        {
+          m_serviceGeodatabase->switchVersionAsync(serviceVersionInfo->name()).then(this, [this]()
```

Comment:
Please refactor this logic. The workflow seems sound, but it can be difficult to follow with so many nested .then sequential blocks.

So rather than (conceptually):
```c++
connect(m_map, &Map::doneLoading, this, [this]() {
  // lots of logic here.
  connect(m_serviceGeodatabase, &ServiceGeodatabase::doneLoading, this, [this, params]() {
    // lots of logic here.
    m_serviceGeodatabase->createVersionAsync().then({
      // lots of logic here.
      m_serviceGeodatabase->switchVersionAsync().then({
      // lots of logic here.
```
That is perfectly fine if each block is simple, but these are not simple. It becomes hard to follow the logic.

Consider having separate `onDoSomething` handlers that you call. That should help to break up the workflow into smaller, readable chunks.

```c++
connect(m_map, &Map::doneLoading, this, &ValidateUtilityNetworkTopology::onMapLoaded);

//...

void ValidateUtilityNetworkTopology::onMapLoaded()
{
  // do stuff here.
}

void ValidateUtilityNetworkTopology::onServiceGeodatabaseLoaded()
{
  m_serviceGeodatabase->createVersionAsync(params).then(this, [this](ServiceVersionInfo* serviceVersionInfo){
    m_serviceGeodatabase->switchVersionAsync(serviceVersionInfo->name()).then(this, [this]()
    {
      onSwitchVersionCompleted();
      // ...
// ... and so on
```

If you have simple logic, then you can keep those lambdas inline like you have it.

Reply:
separated into different methods